tests/mem_tracking.rs

```diff
@@ -0,0 +1,158 @@
+// Copyright (C) Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use core::fmt::Debug;
+use parity_scale_codec::{
+	alloc::{
+		collections::{BTreeMap, BTreeSet, LinkedList, VecDeque},
+		rc::Rc,
+	},
+	DecodeWithMemTracking, Encode, Error, MemTrackingInput,
+};
+use parity_scale_codec_derive::{Decode as DeriveDecode, Encode as DeriveEncode};
+
+const ARRAY: [u8; 1000] = [11; 1000];
+
+#[derive(DeriveEncode, DeriveDecode, PartialEq, Debug)]
+#[allow(clippy::large_enum_variant)]
+enum TestEnum {
+	Empty,
+	Array([u8; 1000]),
+}
+
+impl DecodeWithMemTracking for TestEnum {}
+
+#[derive(DeriveEncode, DeriveDecode, PartialEq, Debug)]
+struct ComplexStruct {
+	test_enum: TestEnum,
+	boxed_test_enum: Box<TestEnum>,
+	box_field: Box<u32>,
+	vec: Vec<u8>,
+}
+
+impl DecodeWithMemTracking for ComplexStruct {}
+
+fn decode_object<T>(obj: T, mem_limit: usize, expected_used_mem: usize) -> Result<T, Error>
```

Comment:
This function could also call the decode again with one byte less memory limit to check that it fails.  
Then you get all the error cases also covered.